exec/java-exec/src/main/java/org/apache/drill/exec/store/log/LogRecordReader.java

```diff
@@ -0,0 +1,261 @@
+    } else {
+      //TODO Check for invalid regex
+      r = Pattern.compile(regex);
+      Matcher m = r.matcher("test");
+      if (m.groupCount() == 0) {
+        throw UserException.parseError().message("Invalid Regular Expression: No Capturing Groups", 0).build(logger);
+      } else if (m.groupCount() != (fieldNames.size())) {
+        throw UserException.parseError().message("Invalid Regular Expression: Field names do not match capturing groups.  There are " + m.groupCount() + " captured groups in the data and " + fieldNames.size() + " specified in the configuration.", 0).build(logger);
+
+      } else if ((dataTypes == null) || m.groupCount() != dataTypes.size()) {
+        //If the number of data types is not correct, create a list of varchar
+        dataTypes = new ArrayList<String>();
+        for (int i = -0; i < m.groupCount(); i++) {
+          dataTypes.add("VARCHAR");
+        }
+      }
+    }
+
+    //Check and set up date formats
+    if (dataTypes.contains("DATE") || dataTypes.contains("TIMESTAMP")) {
+      if (dateFormat != null && !dateFormat.isEmpty()) {
+        df = new java.text.SimpleDateFormat(dateFormat);
+      } else {
+        throw UserException.parseError().message("Invalid date format.  The date formatting string was empty.  Please specify a valid date format string in the configuration for this data source.", 0).build(logger);
+      }
+    }
+
+    if (dataTypes.contains("TIME")) {
+      if (timeFormat != null && !timeFormat.isEmpty()) {
+        tf = new java.text.SimpleDateFormat(timeFormat);
+      } else {
+        throw UserException.parseError().message("Invalid time format.  The time formatting string was empty.  Please specify a valid time format string in the configuration for this data source.", 0).build(logger);
+      }
+    }
+
+  }
+
+  public int next() {
+    this.writer.allocate();
+    this.writer.reset();
+
+    int recordCount = 0;
+
+    try {
+      BaseWriter.MapWriter map = this.writer.rootAsMap();
+      String line = null;
+
+      while (recordCount < MAX_RECORDS_PER_BATCH && (line = this.reader.readLine()) != null) {
+        lineCount++;
+
+        // Skip empty lines
+        if (line.trim().length() == 0) {
+          continue;
+        }
+
+        this.writer.setPosition(recordCount);
+        map.start();
+
+        Matcher m = r.matcher(line);
+        if (m.find()) {
+          for (int i = 1; i <= m.groupCount(); i++) {
+
+            String fieldName = fieldNames.get(i - 1);
+            String type = dataTypes.get(i - 1);
+            String fieldValue;
+
+            fieldValue = m.group(i);
+
+            if (fieldValue == null) {
+              fieldValue = "";
+            }
+
+            if (type.toUpperCase().equals("INT") || type.toUpperCase().equals("INTEGER")) {
+              map.integer(fieldName).writeInt(Integer.parseInt(fieldValue));
+            } else if (type.toUpperCase().equals("DOUBLE") || type.toUpperCase().equals("FLOAT8")) {
```

Comment:
Even better. This is an inner loop. Presumably we want it to go fast (which is why it is deeply nested.)

So, rather than doing (slow) string compares (you could have used a `switch` statement instead), in the `setup()` method, map names to ordinals: `MinorType.ordinal()`. Then do a `switch` on the resulting numbers using, say, `MinorType.VARCHAR.ordinal()`. Much faster.